script/LibStateDiffChecker.sol

```diff
@@ -0,0 +1,154 @@
+        internal
+        pure
+        returns (StateDiffSpec memory stateDiffSpec_)
+    {
+        // Get the chain ID from the first account access.
+        stateDiffSpec_.chainId = accountAccesses[0].chainInfo.chainId;
+
+        // First iterate through the account accesses and storage accesses to count the number of storage modifications,
+        // so that we can allocate the correct amount of memory for the StorageDiffSpec.
+        uint256 modifiedCount = 0;
+        for (uint256 acctIndex = 0; acctIndex < accountAccesses.length; acctIndex++) {
+            for (
+                uint256 storageIndex = 0;
+                storageIndex < accountAccesses[acctIndex].storageAccesses.length;
+                storageIndex++
+            ) {
+                if (
+                    accountAccesses[acctIndex].storageAccesses[storageIndex].previousValue
+                        != accountAccesses[acctIndex].storageAccesses[storageIndex].newValue
+                ) {
+                    modifiedCount++;
+                }
+            }
+        }
+
+        // Allocate memory for the storage modifications.
+        stateDiffSpec_.storageSpecs = new StorageDiffSpec[](modifiedCount);
+
+        // Then iterate again through the account accesses and storage accesses to populate the StorageDiffSpec.
+        uint256 specIndex = 0; // index for next write to stateDiffSpec_.storageSpecs
+        for (uint256 acctIndex = 0; acctIndex < accountAccesses.length; acctIndex++) {
+            for (
+                uint256 storageIndex = 0;
+                storageIndex < accountAccesses[acctIndex].storageAccesses.length;
+                storageIndex++
+            ) {
+                if (
+                    accountAccesses[acctIndex].storageAccesses[storageIndex].previousValue
+                        != accountAccesses[acctIndex].storageAccesses[storageIndex].newValue
+                ) {
+                    stateDiffSpec_.storageSpecs[specIndex] = StorageDiffSpec({
+                        account: accountAccesses[acctIndex].storageAccesses[storageIndex].account,
+                        newValue: accountAccesses[acctIndex].storageAccesses[storageIndex].newValue,
+                        previousValue: accountAccesses[acctIndex].storageAccesses[storageIndex].previousValue,
+                        slot: accountAccesses[acctIndex].storageAccesses[storageIndex].slot
+                    });
+                    specIndex++;
+                }
+            }
+        }
+
+        require(specIndex == modifiedCount, "LibStateDiffChecker: found fewer storage modifications than expected.");
+    }
+
+    /// @notice Checks if a single expected account access matches the actual account access.
+    /// @param expectedDiff The expected account access details.
+    /// @param actualDiff   The actual account access details recorded by the EVM.
+    function checkStateDiff(StateDiffSpec memory expectedDiff, StateDiffSpec memory actualDiff) internal pure {
```

Comment:
Should we add a catch-call `keccak256(abi.encode(expectedDiff)) == keccak256(abi.encode(actualDiff))` check at the end for any fields we forgot to add here?

Reply:
Added this.